### PR TITLE
Allow users to use installed git and option to control credentials on their own [IN PROGRESS]

### PR DESCRIPTION
**Pull request by TingluoHuang**
…able.

### CODE REVIEW

**Comment by TingluoHuang** [comment]
#389

File: src/Agent.Worker/Build/BuildJobExtension.cs [hunk 1]
```diff
@@ -223,6 +222,16 @@ private bool TrySetPrimaryEndpointAndProviderInfo(IExecutionContext executionCon
                 if (SourceProvider != null)
                 {
                     SourceEndpoint = ep;
+
+                    // set endpoint id to repository id
+                    string repoIdString;
+                    Guid repoId;
+                    if (SourceEndpoint.Data.TryGetValue("repositoryId", out repoIdString) &&
+                        Guid.TryParse(repoIdString, out repoId))
```
Comment (ericsciple) [inline src/Agent.Worker/Build/BuildJobExtension.cs]: For TFVC today the variable gets set to "$/", which won't parse as a guid.

Comment (ericsciple) [inline src/Agent.Worker/Build/BuildJobExtension.cs]: "today" as in old Win agent.

Comment (TingluoHuang) [inline src/Agent.Worker/Build/BuildJobExtension.cs]: :D totally forgot TFVC. 
is there any change endpoint ID still matter?
old agent: endpoint id = repoid or $/
new agent: endpoint id = guid.empty.

File: src/Agent.Worker/Build/GitCommandManager.cs [hunk 2]
```diff
@@ -64,47 +65,73 @@ public interface IGitCommandManager : IAgentService
...
-            varUtil.SetEnvironmentVariable("GIT_EXEC_PATH", _gitExecPathEnv);
-
-            // Set GIT_TEMPLATE_DIR.
-            string _gitTemplatePathEnv = Path.Combine(IOUtil.GetExternalsPath(), "git", "share", "git-core", "templates");
-            ArgUtil.Directory(_gitTemplatePathEnv, nameof(_gitTemplatePathEnv));
-            context.Debug($"Set git templateDir to: {_gitTemplatePathEnv}");
-            varUtil.SetEnvironmentVariable("GIT_TEMPLATE_DIR", _gitTemplatePathEnv);
-#endif
+            // Determine do we need to provide creds to git operation
+            bool selfManageGitCreds = context.Variables.GetBoolean("sytem.selfmanagegitcreds") ?? false;
```
Comment (TingluoHuang) [inline src/Agent.Worker/Build/GitCommandManager.cs]: system

File: src/Agent.Worker/Build/GitCommandManager.cs [hunk 3]
```diff
@@ -64,47 +65,73 @@ public interface IGitCommandManager : IAgentService
...
         private string _gitPath = null;
         private Version _version = null;
+        private bool _handleAuthChallenge = true;
+
+        public bool HandleAuthChallenge => _handleAuthChallenge;
 
         public async Task LoadGitExecutionInfo(IExecutionContext context)
         {
             // Resolve the location of git.
+            bool preferGitFromPath = context.Variables.GetBoolean("system.prefergitfrompath") ?? false;
```
Comment (ankitgo) [inline src/Agent.Worker/Build/GitCommandManager.cs]: "system.prefergitfrompath" -- > move to constants?

Comment (TingluoHuang) [inline src/Agent.Worker/Build/GitCommandManager.cs]: I will. :D since your PR merged.

**Comment by bryanmacfarlane** [comment]
How close are we on this?

**Comment by TingluoHuang** [comment]
@bryanmacfarlane i will finish cleanup during weekends and check-in.

**Pull request merged**